File: worksheets/dp-four/kotlin/src/memento/Originator.kt
package memento

data class Originator(
        var x: Double,
        var y: Double,
        var careTaker: CareTaker
) {

    private var lastUndoSavepoint: String = TODO()

    init {
        createSavepoint("INITIAL")
    }

    fun createSavepoint(savepointName: String) {
        TODO()
    }

    fun undo() {
        TODO()
    }

    fun undo(savepointName: String) {
        TODO()
    }

    fun undoAll() {
        TODO()
    }

    private fun setOriginatorState(savepointName: String) {
        TODO()
    }
}

File: worksheets/dp-four/scala/src/memento/Originator.scala
package memento

case class Originator(
  var x: Double,
  var y: Double,
  var careTaker: CareTaker
) {

  private var lastUndoSavepoint: String = _

  createSavepoint("INITIAL")

  def createSavepoint(savepointName: String): Unit = ???

  def undo(): Unit = ???

  def undo(savepointName: String): Unit = ???

  def undoAll(): Unit = ???

  private def setOriginatorState(savepointName: String): Unit = ???

  override def toString(): String = "X: " + x + ", Y: " + y
}
